### PR TITLE
Fix leaking goroutines in QuotaEvaluator

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref https://github.com/kubernetes/kubernetes/issues/108483

```release-note
NONE
```

/kind cleanup
/priority important-longterm
/sig api-machinery

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/controller.go [hunk 1]
```diff
@@ -131,15 +131,28 @@ func NewQuotaEvaluator(quotaAccessor QuotaAccessor, ignoredResources map[schema.
...
 		config:  config,
 	}
+
+	// The queue underneath is starting a goroutine for metrics
+	// exportint that is only stopped on calling ShutDown.
+	// Given that QuotaEvaluator is created for each layer of apiserver
+	// and often not started for some of those (e.g. aggregated apiserver)
+	// we explicitly shut it down on stopCh signal even if it wasn't
+	// effectively started.
+	go evaluator.shutdownOnStop()
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/controller.go:141]: does it means we start only once but we shutdown multiple times?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/controller.go:141]: Sorry - maybe I wasn't clear.
What I meant is that:

- we create a separate QuotaEvaluator for every layer (separate for kube-apiserver, separate for agregated, separate for apiextensions)
- the creation of it actually creates a queue which starts a goroutine that is updating metrics (this is created in line 125 so once per quota evaluator)
- the newly added shutdownOnStop is also created here so called exactly once (per QuotaEvaluator)
- the start() is called *at most* once - line 606

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/controller.go:141]: and why is not valid to keep it as it was?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/controller.go:141]: Not sure I understand the question.

As I wrote above - when I create QuotaEvaluator, this is underneath starting a gorotuine (created by `workqueue.NewNamed`).
That goroutine is shut down only when run() (renamed to start()) method is called.

run() is called on the first `Evaluate` - if that is not called - the goroutine will be leaked.

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/controller.go:141]: ahhhhhhhhhhhhhhhhhhhhhhhh ... sorry friday and lack coffeeeeeee

**Comment by aojea** [comment]
/lgtm
:upside_down_face:

**Pull request merged**